Makefile.cache

```diff
@@ -0,0 +1,637 @@
+#          GIT_CONTENT_SHA  - SHA value is calculated from the target dependency files content.
+#   Cache is loaded from either local cache or global cache based on the dependency SHA match with the cache filename.
+#   Otherwise it builds the package from source.
+# TODO:
+# 1. Extend dpkg for all the chip vendor packages.
+
+# Args:
+# $(1) => target name
+# $(2) => target output file name
+define LOAD_FROM_CACHE
+
+	# Calculate the modules SHA and its dependency SHA value
+	$(call GET_MOD_DEP_SHA,$(1))
+	$(call GET_MOD_SHA,$(1))
+
+	# Form the cache file name
+    $(eval $(1)_MOD_CACHE_FILE      := $(1)-$($(1)_DEP_MOD_SHA)-$($(1)_MOD_HASH).tgz)
+	$(if $(MDEBUG), $(info MODE_CACHE_FILE := $($(1)_MOD_CACHE_FILE)))
+
+	# Retrive and log files list that are modified for the target.
+	$(call GET_MODIFIED_FILES,$(1))
+	$(if $($(1)_FILES_MODIFIED),
+		echo "Target $(1) dependencies are modifed - global cache skipped" >> $($(1)_DST_PATH)/$(1).log
+		echo "Modified dependencies are : [$($(1)_FILES_MODIFIED)] " >> $($(1)_DST_PATH)/$(1).log
+		$(eval $(1)_CACHE_DIR := $(SONIC_DPKG_LOCAL_CACHE_DIR)))
+
+	# Choose the cache file path in the following order
+	# 1. First load from Local cache path
+	# 2. If not, load from global cache path
+	$(eval CACHE_FILE_SELECT:=$(or $(wildcard $(SONIC_DPKG_LOCAL_CACHE_DIR)/$($(1)_MOD_CACHE_FILE)), \
+                                   $(wildcard $(SONIC_DPKG_CACHE_DIR)/$($(1)_MOD_CACHE_FILE))) )
+
+	# Check if any of the derived package is not built
+	$(eval LOAD_DRV_DEB := $(foreach pkg,$(addprefix $($(1)_DST_PATH)/,$(1) $($(1)_DERIVED_DEBS) $($(1)_EXTRA_DEBS)),$(if $(wildcard $(pkg)),,$(pkg))))
+
+	# Load the cache if cache is eanbled and cache file is present in the cache
+	# Update the cache_loaded variable
+	$(if $(and $(CACHE_FILE_SELECT),$(filter cache,$(SONIC_DPKG_CACHE_METHOD))),
+		$(if $(LOAD_DRV_DEB), $($(1)_CACHE_USER) tar -C $($(1)_BASE_PATH) -mxzvf $(CACHE_FILE_SELECT) 1>> $($(1)_DST_PATH)/$(1).log ,echo );
+		echo "File $(CACHE_FILE_SELECT) is loaded from cache" >> $($(1)_DST_PATH)/$(1).log
+		$(eval $(1)_CACHE_LOADED := Yes)
+		$(shell touch $(CACHE_FILE_SELECT))
+		echo "[ CACHE::LOADED ] $($(1)_CACHE_DIR)/$($(1)_MOD_CACHE_FILE)" >> $($(1)_DST_PATH)/$(1).log
+		,
+		echo "File $($(1)_CACHE_DIR)/$($(1)_MOD_CACHE_FILE)  is not present in cache or cache mode set as $(SONIC_DPKG_CACHE_METHOD) !" >> $($(1)_DST_PATH)/$(1).log
+		echo "[ CACHE::SKIPPED ] $($(1)_CACHE_DIR)/$($(1)_MOD_CACHE_FILE)" >> $($(1)_DST_PATH)/$(1).log
+		echo "[ CACHE::SKIPPED ] DEP_FILES - Modified Files: [$($(1)_FILES_MODIFIED)] " >> $($(1)_DST_PATH)/$(1).log
+		echo "[ CACHE::SKIPPED ] DEPENDS   - Modified Files: [$?] " >> $($(1)_DST_PATH)/$(1).log
+	 )
+endef
+
+# Saves the deb package into debian cache
+# A single tared-zip cache is created for .deb and its derived packages in the cache direcory.
+# It saves the .deb into global cache only when its dependencies are not changed,
+#    Otherwise it saves the .deb into local cache
```

Comment:
What is the scenario for the local cache? What is difference between local and global cache? The local cache will be used when the local branch is different from the git remote branch?

Reply:
Local cache is used when there is any modification in the local repo that are yet to be commited in the master repo. This is mainly useful when developer continue to make changes in the local repo. So the cache is created only on the local cache instead of global cache.  This will avoid the global cache update for every incremental change in the local source code. The global cache is shared across multiple users where the local cache is specific to particular repo